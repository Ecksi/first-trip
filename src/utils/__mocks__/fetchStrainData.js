const fetchStrainData = jest.fn().mockImplementation(() => {
  Promise.resolve({
    "Afpak": {
      "id": 1,
      "race": "hybrid",
      "flavors": [
        "Earthy",
        "Chemical",
        "Pine"
      ],
      "effects": {
        "positive": [
          "Relaxed",
          "Hungry",
          "Happy",
          "Sleepy"
        ],
        "negative": [
          "Dizzy"
        ],
        "medical": [
          "Depression",
          "Insomnia",
          "Pain",
          "Stress",
          "Lack of Appetite"
        ]
      }
    }
  });
});

export default fetchStrainData;