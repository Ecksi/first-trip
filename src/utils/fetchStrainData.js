import { strainKey } from '../private/keys';
import { doFetch } from './doFetch';

const fetchStrainData = async () => {
  const url = `http://strainapi.evanbusse.com/${strainKey}/strains/search/all`;
  const strainData = await doFetch(url);

  return cleanStrainData(strainData);
};

const cleanStrainData = dirtyData => {
  const strainNames = Object.keys(dirtyData);
  const strainData = Object.values(dirtyData);
  let cleanStrainData = [];

  for (let i = 0; i < strainData.length; i++) {
    cleanStrainData.push({...strainData[i], name: strainNames[i]});
  }
  
  return cleanStrainData;
};

export default fetchStrainData;
