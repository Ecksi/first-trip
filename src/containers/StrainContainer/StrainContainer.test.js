import React from 'react';
import { shallow } from 'enzyme';
import { StrainContainer, mapStateToProps, mapDispatchToProps } from './index';
import fetchStrainData from '../../utils/fetchStrainData';
import fetchStrainEffects from '../../utils/fetchEffectsData';

jest.mock('../../utils/fetchStrainData');
jest.mock('../../utils/fetchEffectsData');

describe('StrainContainer', () => {
  let strainContainer;
  let mockProps;

  beforeEach(() => {
    mockProps = {
      addStrains: jest.fn(),
      addEffects: jest.fn(),
      name: 'weed',
      strains: [{ name: 'weed' }],
      race: 'sativa',
      effects: {type: {positive: 'happy'}},
      filters: ['happy'],
      results: [{ name: 'weedTwo' }],
      searchResults: jest.fn(),
      searchByFilters: jest.fn(),
      resetFilter: jest.fn()
    };

    strainContainer = shallow(
      <StrainContainer {...mockProps} />,
      { disableLifecycleMethods: true });
  });

  describe('component', () => {
    it('should match the snapshot', () => {
      expect(strainContainer).toMatchSnapshot();
    });
  });

  describe('getStrainData', () => {
    it('calls fetchStrainData', async () => {
      await strainContainer.instance().getStrainData();
      
      expect(fetchStrainData).toHaveBeenCalled();
    });
  });

  describe('getStrainEffects', () => {
    it('calls fetchStrainEffects', async () => {
      await strainContainer.instance().getStrainEffects();

      expect(fetchStrainEffects).toHaveBeenCalledTimes(1);
    });
  });

  describe('mapStateToProps', () => {
    it('can access all strains from the store', () => {
      const strains = mockProps.strains;
      const mockState = { strains };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(mockState);
    });
    
    it('can access all effects from the store', () => {
      const effects = mockProps.effects;
      const mockState = { effects };
      const mappedProps = mapStateToProps(mockState);
      const expected = { type: { positive: 'happy' } };

      expect(mappedProps.effects).toEqual(expected);
    });

    it('can access all filters from the store', () => {
      const filters = mockProps.filters;
      const mockState = { filters };
      const mappedProps = mapStateToProps(mockState);
      const expected = ['happy'];

      expect(mappedProps.filters).toEqual(expected);
    });

    it('can access all results from the store', () => {
      const results = mockProps.results;
      const mockState = { results };
      const mappedProps = mapStateToProps(mockState);
      const expected = [{ name: 'weedTwo' }];

      expect(mappedProps.results).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls addStrains dispatch with the correct params', () => {
      const mockDispatch = jest.fn();
      const strains = [{ name: 'weed' }];
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'ADD_STRAINS',
        strains
      };

      mappedProps.addStrains(strains);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls addEffects dispatch with the correct params', () => {
      const mockDispatch = jest.fn();
      const effects = { type: { positive: 'happy' } };
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'ADD_EFFECTS',
        effects
      };

      mappedProps.addEffects(effects);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls addFilters dispatch with the correct params', () => {
      const mockDispatch = jest.fn();
      const filters = ['happy'];
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'ADD_FILTER',
        filters
      };

      mappedProps.addFilters(filters);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls searchResults dispatch with the correct params', () => {
      const mockDispatch = jest.fn();
      const results = [{ name: 'weedTwo' }];
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'SEARCH_RESULTS',
        results
      };

      mappedProps.searchResults(results);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('calls resetFilter dispatch', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'RESET_FILTER'
      };

      mappedProps.resetFilter();

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    })
  });
});