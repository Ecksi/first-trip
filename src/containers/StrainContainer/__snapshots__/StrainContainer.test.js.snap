// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrainContainer component should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StrainContainer
    addEffects={[Function]}
    addStrains={[Function]}
    effects={
        Object {
            "type": Object {
              "positive": "happy",
            },
          }
    }
    filters={
        Array [
            "happy",
          ]
    }
    name="weed"
    race="sativa"
    resetFilter={[Function]}
    results={
        Array [
            Object {
              "name": "weedTwo",
            },
          ]
    }
    searchByFilters={[Function]}
    searchResults={[Function]}
    strains={
        Array [
            Object {
              "name": "weed",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Hello there 
</h1>,
        <h3>
          Select the effects you would like to have or avoid:
</h3>,
        <div
          className="effects"
>
          <h2>
                    Positive Effects
          </h2>
          <h2>
                    Medical Effects
          </h2>
          <h2>
                    Negative Effects
          </h2>
</div>,
        <div
          className="effects"
>
          <ul />
          <ul />
          <ul />
</div>,
        <div
          className="effects"
>
          <NavLink
                    activeClassName="active"
                    ariaCurrent="true"
                    to="/results"
          >
                    <button
                              className="filter"
                              name="positive"
                              onClick={[Function]}
                    >
                              Filter by Positive
                    </button>
                    <button
                              className="filter"
                              name="medical"
                              onClick={[Function]}
                    >
                              Filter by Medical
                    </button>
                    <button
                              className="filter"
                              name="negative"
                              onClick={[Function]}
                    >
                              Filter by Negative
                    </button>
          </NavLink>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hello there ",
        },
        "ref": null,
        "rendered": "Hello there ",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Select the effects you would like to have or avoid:",
        },
        "ref": null,
        "rendered": "Select the effects you would like to have or avoid:",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Positive Effects
</h2>,
            <h2>
              Medical Effects
</h2>,
            <h2>
              Negative Effects
</h2>,
          ],
          "className": "effects",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Positive Effects",
            },
            "ref": null,
            "rendered": "Positive Effects",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Medical Effects",
            },
            "ref": null,
            "rendered": "Medical Effects",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Negative Effects",
            },
            "ref": null,
            "rendered": "Negative Effects",
            "type": "h2",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ul />,
            <ul />,
            <ul />,
          ],
          "className": "effects",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "ul",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "ul",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "ul",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <NavLink
            activeClassName="active"
            ariaCurrent="true"
            to="/results"
>
            <button
                        className="filter"
                        name="positive"
                        onClick={[Function]}
            >
                        Filter by Positive
            </button>
            <button
                        className="filter"
                        name="medical"
                        onClick={[Function]}
            >
                        Filter by Medical
            </button>
            <button
                        className="filter"
                        name="negative"
                        onClick={[Function]}
            >
                        Filter by Negative
            </button>
</NavLink>,
          "className": "effects",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeClassName": "active",
            "ariaCurrent": "true",
            "children": Array [
              <button
                className="filter"
                name="positive"
                onClick={[Function]}
>
                Filter by Positive
</button>,
              <button
                className="filter"
                name="medical"
                onClick={[Function]}
>
                Filter by Medical
</button>,
              <button
                className="filter"
                name="negative"
                onClick={[Function]}
>
                Filter by Negative
</button>,
            ],
            "to": "/results",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Filter by Positive",
                "className": "filter",
                "name": "positive",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Filter by Positive",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Filter by Medical",
                "className": "filter",
                "name": "medical",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Filter by Medical",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Filter by Negative",
                "className": "filter",
                "name": "negative",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Filter by Negative",
              "type": "button",
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Hello there 
</h1>,
          <h3>
            Select the effects you would like to have or avoid:
</h3>,
          <div
            className="effects"
>
            <h2>
                        Positive Effects
            </h2>
            <h2>
                        Medical Effects
            </h2>
            <h2>
                        Negative Effects
            </h2>
</div>,
          <div
            className="effects"
>
            <ul />
            <ul />
            <ul />
</div>,
          <div
            className="effects"
>
            <NavLink
                        activeClassName="active"
                        ariaCurrent="true"
                        to="/results"
            >
                        <button
                                    className="filter"
                                    name="positive"
                                    onClick={[Function]}
                        >
                                    Filter by Positive
                        </button>
                        <button
                                    className="filter"
                                    name="medical"
                                    onClick={[Function]}
                        >
                                    Filter by Medical
                        </button>
                        <button
                                    className="filter"
                                    name="negative"
                                    onClick={[Function]}
                        >
                                    Filter by Negative
                        </button>
            </NavLink>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hello there ",
          },
          "ref": null,
          "rendered": "Hello there ",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Select the effects you would like to have or avoid:",
          },
          "ref": null,
          "rendered": "Select the effects you would like to have or avoid:",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Positive Effects
</h2>,
              <h2>
                Medical Effects
</h2>,
              <h2>
                Negative Effects
</h2>,
            ],
            "className": "effects",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Positive Effects",
              },
              "ref": null,
              "rendered": "Positive Effects",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Medical Effects",
              },
              "ref": null,
              "rendered": "Medical Effects",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Negative Effects",
              },
              "ref": null,
              "rendered": "Negative Effects",
              "type": "h2",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ul />,
              <ul />,
              <ul />,
            ],
            "className": "effects",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "ul",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "ul",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "ul",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <NavLink
              activeClassName="active"
              ariaCurrent="true"
              to="/results"
>
              <button
                            className="filter"
                            name="positive"
                            onClick={[Function]}
              >
                            Filter by Positive
              </button>
              <button
                            className="filter"
                            name="medical"
                            onClick={[Function]}
              >
                            Filter by Medical
              </button>
              <button
                            className="filter"
                            name="negative"
                            onClick={[Function]}
              >
                            Filter by Negative
              </button>
</NavLink>,
            "className": "effects",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeClassName": "active",
              "ariaCurrent": "true",
              "children": Array [
                <button
                  className="filter"
                  name="positive"
                  onClick={[Function]}
>
                  Filter by Positive
</button>,
                <button
                  className="filter"
                  name="medical"
                  onClick={[Function]}
>
                  Filter by Medical
</button>,
                <button
                  className="filter"
                  name="negative"
                  onClick={[Function]}
>
                  Filter by Negative
</button>,
              ],
              "to": "/results",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Filter by Positive",
                  "className": "filter",
                  "name": "positive",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Filter by Positive",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Filter by Medical",
                  "className": "filter",
                  "name": "medical",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Filter by Medical",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Filter by Negative",
                  "className": "filter",
                  "name": "negative",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Filter by Negative",
                "type": "button",
              },
            ],
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
